worker/remoterelations/remoterelationsworker.go

```diff
@@ -0,0 +1,108 @@
+// Copyright 2017 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package remoterelations
+
+import (
+	"github.com/juju/errors"
+	"gopkg.in/juju/names.v2"
+	worker "gopkg.in/juju/worker.v1"
+
+	"github.com/juju/juju/apiserver/params"
+	"github.com/juju/juju/watcher"
+	"github.com/juju/juju/worker/catacomb"
+)
+
+// remoteRelationsWorker listens for changes to the
+// life and status of a relation in the offering model.
+type remoteRelationsWorker struct {
+	catacomb catacomb.Catacomb
+
+	relationTag         names.RelationTag
+	remoteRelationToken string
+	applicationToken    string
+	relationsWatcher    watcher.RelationStatusWatcher
+	changes             chan<- params.RemoteRelationChangeEvent
+}
+
+func newRemoteRelationsWorker(
+	relationTag names.RelationTag,
+	applicationToken string,
+	remoteRelationToken string,
+	relationsWatcher watcher.RelationStatusWatcher,
+	changes chan<- params.RemoteRelationChangeEvent,
+) (*remoteRelationsWorker, error) {
+	w := &remoteRelationsWorker{
+		relationsWatcher:    relationsWatcher,
+		relationTag:         relationTag,
+		remoteRelationToken: remoteRelationToken,
+		applicationToken:    applicationToken,
+		changes:             changes,
+	}
+	err := catacomb.Invoke(catacomb.Plan{
+		Site: &w.catacomb,
+		Work: w.loop,
+		Init: []worker.Worker{relationsWatcher},
+	})
+	return w, err
+}
+
+// Kill is defined on worker.Worker
+func (w *remoteRelationsWorker) Kill() {
+	w.catacomb.Kill(nil)
+}
+
+// Wait is defined on worker.Worker
+func (w *remoteRelationsWorker) Wait() error {
+	return w.catacomb.Wait()
+}
+
+func (w *remoteRelationsWorker) loop() error {
+	var (
+		changes chan<- params.RemoteRelationChangeEvent
+		event   params.RemoteRelationChangeEvent
+	)
+	for {
+		select {
+		case <-w.catacomb.Dying():
+			return w.catacomb.ErrDying()
+		case relChanges, ok := <-w.relationsWatcher.Changes():
+			if !ok {
+				// We are dying.
```

Comment:
can you please make this `return w.catacomb.ErrDying()`? then if there's a bug with the channel being closed, this loop won't spin forever